pkg/remote/codec/unknown/unknown.go

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2021 CloudWeGo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package unknown
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	thrif "github.com/apache/thrift/lib/go/thrift"
+	"github.com/cloudwego/dynamicgo/thrift"
+	"github.com/cloudwego/kitex/pkg/protocol/bthrift"
+	"github.com/cloudwego/kitex/pkg/remote"
+	"github.com/cloudwego/kitex/pkg/remote/codec"
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	unknowns "github.com/cloudwego/kitex/pkg/remote/codec/unknown/service"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/cloudwego/kitex/pkg/serviceinfo"
+)
+
+// UnknownCodec implements PayloadCodec
+type unknownCodec struct {
+	Codec remote.PayloadCodec
+}
+
+// NewUnknownCodec creates the unknown binary codec.
+func NewUnknownCodec(code remote.PayloadCodec) remote.PayloadCodec {
+	return &unknownCodec{code}
+}
+
+// Marshal implements the remote.PayloadCodec interface.
+func (c unknownCodec) Marshal(ctx context.Context, msg remote.Message, out remote.ByteBuffer) error {
+	ink := msg.RPCInfo().Invocation()
+	data := msg.Data()
+
+	res, ok := data.(*unknowns.Result)
+	if !ok {
+		return c.Codec.Marshal(ctx, msg, out)
+	}
+	if len(res.Success) == 0 {
+		return errors.New("unknown messages cannot be empty")
+	}
+	if msg.MessageType() == remote.Exception {
+		return c.Codec.Marshal(ctx, msg, out)
+	}
+	if ink, ok := ink.(rpcinfo.InvocationSetter); ok {
+		ink.SetMethodName(res.Method)
+		ink.SetServiceName(res.ServiceName)
+	} else {
+		return errors.New("the interface Invocation doesn't implement InvocationSetter")
+	}
+	if err := encode(res, msg, out); err != nil {
+		return c.Codec.Marshal(ctx, msg, out)
+	}
+	return nil
+}
+
+// Unmarshal implements the remote.PayloadCodec interface.
+func (c unknownCodec) Unmarshal(ctx context.Context, message remote.Message, in remote.ByteBuffer) error {
+	ink := message.RPCInfo().Invocation()
+	service, method, size, err := decode(message, in)
+	if err != nil {
+		return c.Codec.Unmarshal(ctx, message, in)
+	}
+	err = codec.SetOrCheckMethodName(method, message)
+	if te, ok := err.(*remote.TransError); ok && te.TypeID() == remote.UnknownMethod {
```

Comment:
both UnknownMethod and UnknownService error should be handled

Reply:
ok